src/TinyHelpers/Threading/AsyncLock.cs

```diff
@@ -12,9 +12,40 @@ public class AsyncLock : IDisposable
     /// </summary>
     /// <param name="cancellationToken">A token that can be used to request cancellation of the asynchronous operation.</param>
     /// <returns>An awaitable task.</returns>
-    public async Task<AsyncLock> LockAsync(CancellationToken cancellationToken = default)
+    public Task<AsyncLock> LockAsync(CancellationToken cancellationToken = default)
     {
-        await semaphoreSlim.WaitAsync(cancellationToken).ConfigureAwait(false);
+        return LockAsync(Timeout.Infinite, cancellationToken);
+    }
+
+    /// <summary>
+    /// Asyncronously waits for the lock to become available, with a specific timeout in milliseconds.
+    /// </summary>
+    /// <param name="timeoutInMilliseconds">The number of milliseconds to wait, or <see cref="Timeout.Infinite"/>(-1) to wait indefinitely.</param>
+    /// <param name="cancellationToken">A token that can be used to request cancellation of the asynchronous operation.</param>
+    /// <exception cref="ArgumentOutOfRangeException">
+    /// <paramref name="timeoutInMilliseconds"/> is a negative number other than -1 milliseconds, which represents
+    /// an infinite timeout.
+    /// </exception>
+    /// <returns>An awaitable task.</returns>
+    public async Task<AsyncLock> LockAsync(int timeoutInMilliseconds, CancellationToken cancellationToken = default)
+    {
+        await semaphoreSlim.WaitAsync(timeoutInMilliseconds, cancellationToken).ConfigureAwait(false);
+        return this;
+    }
```

Comment:
This solution isn't correct, because the `SemaphoreSlim.WaitAsync` methods that take a time interval, return a boolean to inform if the current thread successfully entered the semaphore (see for example https://learn.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim.waitasync?view=net-9.0#system-threading-semaphoreslim-waitasync(system-int32-system-threading-cancellationtoken).

This information must be propagated to the caller, otherwise it may incorrectly think to have acquired the lock, causing issues. For example, this method could return `null` if the lock hasn't been acquired. Or a struct with two properties: `IsOwned` and `AsyncLock?`, where the first tells if the lock is owned (and in this case, the second is the `AsyncLock` itself). If you decide to follow this way (that probably is the more correct), I suggest you to define also a _Deconstruct_ in the tuple to make its usage more straightforward, like this: https://github.com/marcominerva/TinyHelpers/blob/ca6602cd5ab8abf3be9a1388e4624a4f80c32a83/src/TinyHelpers/Models/WithIndex.cs#L31-L37